pydough/errors/pydough_error_builder.py

```diff
@@ -0,0 +1,344 @@
+            node._parcel[0], UnqualifiedRoot
+        ):
+            suggestions: list[str] = find_possible_name_matches(
+                term_name=node._parcel[1],
+                candidates=set(node._parcel[0]._parcel[1]),
+                atol=2,
+                rtol=0.1,
+                min_names=3,
+                insert_cost=0.5,
+                delete_cost=1.0,
+                substitution_cost=1.0,
+                capital_cost=0.1,
+            )
+
+            # Check if there are any suggestions to add
+            if len(suggestions) > 0:
+                suggestions_str: str = ", ".join(suggestions)
+                error_message += f" Did you mean: {suggestions_str}?"
+        else:
+            error_message += " Did you mean to use a function?"
+        return PyDoughUnqualifiedException(error_message)
+
+    def bad_window_per(
+        self,
+        per: str,
+        ancestral_names: list[str],
+        context: "PyDoughCollectionQDAG",
+        window: "UnqualifiedWindow",
+    ) -> PyDoughException:
+        """
+        Creates an exception for when the `per` string in a window is malformed.
+
+        Args:
+            `per`: The per string that caused the error.
+            `ancestral_names`: The names of the ancestors in the context.
+            `context`: The collection context where the error occurred.
+            `window`: The unqualified window that contains the per string.
+
+        Returns:
+            An exception indicating the malformed per string.
+        """
+        ancestor_name: str
+        ancestor_idx: int | None = None
+        msg: str | None = None
+        components: list[str] = per.split(":")
+
+        # Extract the name/idx components of `per=name:idx`, identifying an
+        # error if not in that format.
+        if len(components) <= 2:
+            if len(components) == 1:
+                ancestor_name = components[0]
+                ancestor_idx = None
+            elif len(components) == 2:
+                ancestor_name = components[0]
+                if not components[1].isdigit() or int(components[1]) <= 0:
+                    msg = "expected the index after ':' to be a positive integer"
+                else:
+                    ancestor_idx = int(components[1])
+            # If an error was not found yet, figure out what is wrong with
+            # `name` or `idx`.
+            if msg is None:
+                if ancestor_name not in ancestral_names:
+                    msg = f"unrecognized ancestor {ancestor_name!r}"
+                elif ancestor_idx is None and ancestral_names.count(ancestor_name) > 1:
+                    msg = f"per-string {ancestor_name!r} is ambiguous in this context; use the form '{ancestor_name}:index' to disambiguate, where '{ancestor_name}:1' refers to the most recent ancestor"
+                else:
+                    msg = f"there are not {ancestor_idx} ancestors of the current context with name {ancestor_name!r}"
```

Comment:
This will run if `ancestor_idx is None` and `ancestral_names.count(ancestor_name)` is `<= 1`
If that's intentional, then `msg` should be updated as currently it'll have `there are not {ancestor_idx} ancestors` as that will be `there are not None ancestors`
If not intentional, then `else` should be `elif ancestor_idx is not None`